src/twisted/mail/smtp.py

```diff
@@ -1922,21 +1927,37 @@ def clientConnectionLost(self, connector, err):
         self._processConnectionError(connector, err)
 
     def _processConnectionError(self, connector, err):
+        # If the initial connection succeeded, but there was a later error
+        # (such as TLS validation failure), we're more interested in that than
+        # "Connection was closed cleanly" or whatever.
+        if (
+            self.currentProtocol is not None
+            and self.currentProtocol.failureReason is not None
+        ):
+            err = self.currentProtocol.failureReason
+
         self.currentProtocol = None
-        if (self.retries < 0) and (not self.sendFinished):
+        if self.sendFinished:
+            # we already completed the send, and the deferred has been called.
+            # There is nothing more we can do.
+            return
+
+        if self.retries < 0:
             log.msg("SMTP Client retrying server. Retry: %s" % -self.retries)
 
             # Rewind the file in case part of it was read while attempting to
             # send the message.
             self.file.seek(0, 0)
             connector.connect()
             self.retries += 1
-        elif not self.sendFinished:
-            # If we were unable to communicate with the SMTP server a ConnectionDone will be
-            # returned. We want a more clear error message for debugging
-            if err.check(error.ConnectionDone):
-                err.value = SMTPConnectError(-1, "Unable to connect to server.")
-            self.result.errback(err.value)
+            return
+
+        # do our best to wrap the error into something more meaningful.
+        exn = SMTPConnectError(
+            -1, "Unable to connect to SMTP server: " + str(err.value)
+        )
+        exn.__cause__ = err.value
```

Comment:
I appreciate the use of `__cause__` to chain the exception properly; we should probably make this a bit easier to do with `Failure`.